Checks/PWR079/README.md

```diff
@@ -0,0 +1,336 @@
+$ ./example_array_gcc 
+Sum is: 1.680000
+```
+
+- However, with `clang -O2`, `sum` appears to contain arbitrary data, leading
+  to incorrect results:
+
+```txt
+$ clang --version
+Debian clang version 19.1.5 (1)
+$ clang -O2 example_array.c -o example_array_clang
+$ ./example_array_clang 
+Sum is: nan
+```
+
+The solution is straightforward, always initialize variables before using them:
+
+```c
+double sum = 0.0;
+```
+
+This principle applies to all variable types, including other elemental types
+like `int`, `struct` elements, and pointers.
+
+Pointers are particularly important in C, as they are commonly used to
+represent n-dimensional arrays. However, uninitialized pointers can easily lead
+to invalid memory accesses and program crashes. Let's consider another example
+code with pointers, where a computational function ensures the received
+pointers are valid before accessing their contents:
+
+```c
+// example_matrix.c
+#include <stdio.h>
+
+void perform_computation(double *matrix_a, double *matrix_b) {
+  if (matrix_a == NULL || matrix_b == NULL) {
+    printf("A matrix is NULL; skipping computation\n");
+    return;
+  }
+
+  printf("Performing computation...\n");
+}
+
+int main() {
+  double *matrix_a, *matrix_b;
+  perform_computation(matrix_a, matrix_b);
+
+  return 0;
+}
+```
+
+Note how `matrix_a` and `matrix_b` are never initialized. Although this example
+may seem trivial, similar scenarios can occur in large, complex codebases where
+pointers traverse multiple functions and conditional logic, making such issues
+hard to diagnose and correct.
+
+Since the contents of the pointers are indeterminate, we can obtain different
+results depending on the compiler and settings:
+
+- `gcc -O2` appears to set the pointers to `NULL`, preventing the computation:
+
+```txt
+$ gcc -O2 example_matrix.c -o example_matrix_gcc
+$ ./example_matrix_gcc 
+A matrix is NULL; skipping computation
+```
+
+- However, with `clang -O2`, the pointers seem to hold arbitrary values,
+  allowing the computation to proceed and likely crash due to invalid memory
+  accesses later on:
+
+```txt
+$ clang -O2 example_matrix.c -o example_matrix_clang
+$ ./example_matrix_clang 
+Performing computation...
+```
+
+To help prevent these types of issues, it's a good practice to initialize
+pointers to `NULL` by default if they aren't assigned at their declaration:
+
+```c
+double *matrix_a = NULL, *matrix_b = NULL;
+```
+
+#### Fortran
+
+Consider the following code, which aims to sum the elements of an array:
+
+```fortran
+! example_array.f90
+program main
+  use iso_fortran_env, only: real32
+  implicit none
+
+  real(kind=real32) :: array(5)
+  real(kind=real32) :: sum
+
+  array = [0.24, 0.33, 0.17, 0.89, 0.05]
+  sum = sum_array(array)
+
+  print *, "Sum is:", sum
+
+contains
+
+  pure real(kind=real32) function sum_array(array)
+    implicit none
+    real(kind=real32), intent(in) :: array(:)
+    real(kind=real32) :: sum
```

Comment:
Again, consider highlighting this line in the code for clarity!

Reply:
Done!